Instant query with no data causes a panic

Looks like running an instant query against an empty series set series can cause a panic
```
panic: runtime error: index out of range [4] with length 0

goroutine 348 [running]:
github.com/thanos-community/promql-engine/physicalplan/aggregate.(*aggregate).workerTask(0x100e48a40?, 0x14000091668?, {0x0, {0x0, 0x0, 0x0}, {0x0, 0x0, 0x0}})
	/Users/fpetkovski/Projects/promql-engine/physicalplan/aggregate/hashaggregate.go:155 +0x11c
github.com/thanos-community/promql-engine/worker.(*Worker).Start(0x14000eacc80)
	/Users/fpetkovski/Projects/promql-engine/worker/worker.go:65 +0xdc
created by github.com/thanos-community/promql-engine/worker.Group.Start
	/Users/fpetkovski/Projects/promql-engine/worker/worker.go:24 +0x30
```

--- a/engine/engine_test.go
+++ b/engine/engine_test.go
@@ -261,6 +261,23 @@ func TestQueriesAgainstOldEngine(t *testing.T) {
 				foo{method="get", code="404"} 1+2.2x20`,
 			query: `1 - (100 * sum(foo{method="get"}) / sum(foo))`,
 		},
+		{
+			name:  "empty series",
+			load:  "",
+			query: "http_requests_total",
+		},
+		{
+			name:  "empty series with func",
+			load:  "",
+			query: "sum(http_requests_total)",
+		},
+		{
+			name: "empty result",
+			load: `load 30s
+					http_requests_total{pod="nginx-1"} 1+1x15
+					http_requests_total{pod="nginx-2"} 1+2x18`,
+			query: `http_requests_total{pod="nginx-3"}`,
+		},
 	}
 
 	for _, tc := range cases {
@@ -424,6 +441,23 @@ func TestInstantQuery(t *testing.T) {
 				foo{method="get", code="500"} 1+1.1x30
 				foo{method="get", code="404"} 1+2.2x20`,
 			query: `1 - (100 * sum(foo{method="get"}) / sum(foo))`,
+		},
+		{
+			name:  "empty series",
+			load:  "",
+			query: "http_requests_total",
+		},
+		{
+			name:  "empty series with func",
+			load:  "",
+			query: "sum(http_requests_total)",
+		},
+		{
+			name: "empty result",
+			load: `load 30s
+					http_requests_total{pod="nginx-1"} 1+1x15
+					http_requests_total{pod="nginx-2"} 1+2x18`,
+			query: `http_requests_total{pod="nginx-3"}`,
 		},
 	}
 

--- a/engine/instant_query.go
+++ b/engine/instant_query.go
@@ -43,7 +43,7 @@ func (q *instantQuery) Exec(ctx context.Context) *promql.Result {
 	}
 
 	if len(vs) == 0 {
-		return &promql.Result{}
+		return &promql.Result{Value: promql.Vector{}}
 	}
 
 	resultSeries, err := q.plan.Series(ctx)

--- a/physicalplan/binary/scalar.go
+++ b/physicalplan/binary/scalar.go
@@ -1,3 +1,6 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
 package binary
 
 import (

--- a/physicalplan/binary/table.go
+++ b/physicalplan/binary/table.go
@@ -1,3 +1,6 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
 package binary
 
 import (

--- a/physicalplan/binary/vector.go
+++ b/physicalplan/binary/vector.go
@@ -1,3 +1,6 @@
+// Copyright (c) The Thanos Community Authors.
+// Licensed under the Apache License 2.0.
+
 package binary
 
 import (

--- a/worker/worker.go
+++ b/worker/worker.go
@@ -36,6 +36,8 @@ type Worker struct {
 	input    model.StepVector
 	output   model.StepVector
 
+	// next and done help in starting/stopping Workers
+	// and tracking completion.
 	next *sync.Mutex
 	done *sync.Mutex
 
@@ -46,6 +48,7 @@ type Worker struct {
 type Task func(workerID int, in model.StepVector) model.StepVector
 
 func New(workerID int, task Task) *Worker {
+	// next and done are both locked when a new Worker is spawned.
 	next := &sync.Mutex{}
 	next.Lock()
 	done := &sync.Mutex{}
@@ -61,18 +64,30 @@ func New(workerID int, task Task) *Worker {
 
 func (w *Worker) Start() {
 	for {
+		// Wait for next to be unlocked by Send
+		// before starting work by acquiring lock.
 		w.next.Lock()
-		w.output = w.doWork(w.workerID, w.input)
-		w.done.Unlock()
+		// Shutdown can also unlock next, so in this case
+		// do not run doWork.
+		if !w.exit {
+			w.output = w.doWork(w.workerID, w.input)
+			w.done.Unlock()
+		}
 	}
 }
 
 func (w *Worker) Send(input model.StepVector) {
+	// input is received for Worker, so run Start
+	// by unlocking next.
 	w.input = input
 	w.next.Unlock()
 }
 
 func (w *Worker) Done() {
+	// We can only acquire done lock once Start
+	// has finished executing doWork, i.e the Worker
+	// has finished its task.
+	// If this function returns, it indicates completion.
 	w.done.Lock()
 }
 

